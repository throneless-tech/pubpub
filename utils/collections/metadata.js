import metadataSchemas, { getSchemaForKind } from './schemas';

const isNullOrUndefined = (val) => val === null || val === undefined;

export const mapMetadataFields = (kind, fn) => {
	const schema = getSchemaForKind(kind);
	const res = {};
	schema.metadata.forEach((field) => {
		const result = fn(field);
		if (typeof result !== 'undefined') {
			res[field.name] = result;
		}
	});
	return res;
};

export const getAllSchemaKinds = () => metadataSchemas.map((s) => s.kind);

export const normalizeMetadataToKind = (metadata, kind, context) =>
	mapMetadataFields(kind, (field) => {
		const { name, derivedFrom, defaultDerivedFrom } = field;
		if (derivedFrom) {
			const derived = derivedFrom(context);
			if (!isNullOrUndefined(derived)) {
				return derived;
			}
		}
		const existingValue = metadata[name];
		if (!isNullOrUndefined(existingValue)) {
			return existingValue;
		}
		if (defaultDerivedFrom) {
			return defaultDerivedFrom(context);
		}
		return undefined;
	});

export const deserializeMetadata = ({ metadata, kind, fallback }) =>
	mapMetadataFields(kind, (field) => {
		const { name, type } = field;
		if (metadata[name]) {
			if (type) {
				return type.deserialize(metadata[name]);
			}
			return metadata[name];
		}
		if (fallback) {
			return fallback(field);
		}
		return undefined;
	});

export const enumerateMetadataFields = (metadata, kind) => {
	const { metadata: shape } = getSchemaForKind(kind);
	return shape.map((field) => {
		const { name } = field;
		return {
			...field,
			value: metadata[name],
		};
	});
};
