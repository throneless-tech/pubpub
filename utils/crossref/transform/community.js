import { communityUrl } from 'utils/canonicalUrls';

const getLanguageForCommunity = () => 'en';

export default ({ globals }) => (community) => {
	const { title, issn: denormalizedIssn } = community;
	const issn = denormalizedIssn && denormalizedIssn.replace('-', '');
	return {
		title: title,
		issn: issn,
		timestamp: globals.timestamp,
		language: getLanguageForCommunity(community),
		doi: globals.dois.community,
		url: communityUrl(community),
	};
};
