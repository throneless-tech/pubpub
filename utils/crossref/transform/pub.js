import { pubUrl } from 'utils/canonicalUrls';
import { getPubPublishedDate } from 'utils/pub/pubDates';

import transformAttributions from './attributions';

export default ({ globals, community }) => (pub) => {
	const { timestamp, dois } = globals;
	const { title } = pub;
	const publicationDate = getPubPublishedDate(pub);
	return {
		title: title,
		timestamp: timestamp,
		publicationDate: publicationDate,
		attributions: transformAttributions(pub.attributions),
		resourceUrl: pubUrl(community, pub),
		doi: dois.pub,
	};
};
