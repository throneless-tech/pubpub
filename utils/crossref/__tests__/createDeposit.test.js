/* global describe, it, expect */
import { mapMetadataFields } from '../../collections/metadata';
import { getSchemaForKind } from '../../collections/schemas';

import createDepositPartial from '../createDeposit';

import community from './data/community';
import pub from './data/pub';
import makeCollectionPub from './data/makeCollectionPub';
import { book, issue, conference } from './data/collections';

const createDeposit = (context, target) =>
	createDepositPartial(
		{ ...context, community: community },
		target,
		new Date('2019-04-11T19:02:15.577Z'),
	);

const stripFieldsFromCollectionMetadata = (where) => (collection) => {
	return {
		...collection,
		metadata: mapMetadataFields(collection.kind, (field) => {
			if (where(field)) {
				return undefined;
			}
			return collection.metadata[field.name];
		}),
	};
};

const stripDatesFromCollectionMetadata = stripFieldsFromCollectionMetadata(
	(field) => field.type && field.type.name === 'date',
);

describe('createDeposit', () => {
	it('creates a deposit for a pub', () => {
		expect(
			createDeposit(
				{
					pub: pub,
				},
				'pub',
			),
		).toMatchSnapshot();
	});

	it('creates a deposit for a pub in a book', () => {
		expect(
			createDeposit(
				{
					collection: book,
					collectionPub: makeCollectionPub(book, pub),
					pub: pub,
				},
				'pub',
			),
		).toMatchSnapshot();
	});

	it('creates a deposit for a pub in an issue', () => {
		expect(
			createDeposit(
				{
					collection: issue,
					collectionPub: makeCollectionPub(issue, pub),
					pub: pub,
				},
				'pub',
			),
		).toMatchSnapshot();
	});

	it('creates a deposit for a pub in a conference', () => {
		expect(
			createDeposit(
				{
					collection: conference,
					collectionPub: makeCollectionPub(conference, pub),
					pub: pub,
				},
				'pub',
			),
		).toMatchSnapshot();
	});

	it('creates a deposit for a book', () => {
		expect(
			createDeposit(
				{
					collection: book,
				},
				'collection',
			),
		).toMatchSnapshot();
	});

	it('creates a deposit for an issue', () => {
		expect(
			createDeposit(
				{
					collection: issue,
				},
				'collection',
			),
		).toMatchSnapshot();
	});

	it('creates a deposit for a conference', () => {
		expect(
			createDeposit(
				{
					collection: conference,
				},
				'collection',
			),
		).toMatchSnapshot();
	});

	it('creates a deposit for a book without any date information', () => {
		expect(
			createDeposit(
				{
					collection: stripDatesFromCollectionMetadata(book),
				},
				'collection',
			),
		).toMatchSnapshot();
	});

	it('creates a deposit for an issue without any date information', () => {
		expect(
			createDeposit(
				{
					collection: stripDatesFromCollectionMetadata(issue),
				},
				'collection',
			),
		).toMatchSnapshot();
	});

	it('creates a deposit for a conference without any date information', () => {
		expect(
			createDeposit(
				{
					collection: stripDatesFromCollectionMetadata(conference),
				},
				'collection',
			),
		).toMatchSnapshot();
	});

	it('respects the `url` and `doi` metadata fields for a collection', () => {
		const { deposit } = createDeposit(
			{
				collection: book,
				collectionPub: makeCollectionPub(book, pub),
				pub: pub,
			},
			'pub',
		);
		const {
			doi_data: { doi, resource },
		} = deposit.doi_batch.body.book.book_metadata;
		expect(doi).toEqual('an_utterly_fake_doi');
		expect(resource).toEqual('https://test.com');
	});

	it('correctly handles date metadata provided by a collection', () => {
		const { deposit } = createDeposit(
			{
				collection: conference,
				collectionPub: makeCollectionPub(conference, pub),
				pub: pub,
			},
			'pub',
		);
		const {
			proceedings_metadata: { publication_date },
			event_metadata: { conference_date },
		} = deposit.doi_batch.body.conference;
		expect(conference_date).toEqual({
			'#text': '2019-05-01',
		});
		expect(publication_date).toEqual({
			'@media_type': 'online',
			day: '01',
			month: '05',
			year: '2019',
		});
	});

	it('understands contextHints provided by a CollectionPub', () => {
		const contextHintValue = 'glossary';
		const { deposit } = createDeposit(
			{
				collection: book,
				collectionPub: makeCollectionPub(book, pub, contextHintValue),
				pub: pub,
			},
			'pub',
		);
		const {
			content_item: { '@component_type': componentType },
		} = deposit.doi_batch.body.book;
		expect(componentType).toEqual(
			getSchemaForKind('book').contextHints.find((ch) => ch.value === contextHintValue)
				.crossrefComponentType,
		);
	});
});
