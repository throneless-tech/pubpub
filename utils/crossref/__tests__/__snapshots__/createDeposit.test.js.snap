// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDeposit creates a deposit for a book 1`] = `
Object {
  "deposit": Object {
    "doi_batch": Object {
      "@version": "4.4.1",
      "@xmlns": "http://www.crossref.org/schema/4.4.1",
      "@xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
      "@xsi:schemaLocation": "http://www.crossref.org/schema/4.4.1 http://www.crossref.org/schema/deposit/crossref4.4.1.xsd",
      "body": Object {
        "book": Object {
          "@book_type": "other",
          "book_metadata": Object {
            "contributors": Object {
              "person_name": Array [
                Object {
                  "@contributor_role": "author",
                  "@sequence": "first",
                  "affiliation": "Massachusetts Institute of Technology",
                  "given_name": "Travis",
                  "surname": "Rich",
                },
                Object {
                  "@contributor_role": "reader",
                  "@sequence": "additional",
                  "given_name": "Gabriel",
                  "surname": "Stein",
                },
                Object {
                  "@contributor_role": "reader",
                  "@sequence": "additional",
                  "given_name": "Catherine",
                  "surname": "Ahearn",
                },
                Object {
                  "@contributor_role": "author",
                  "@sequence": "additional",
                  "affiliation": "Springfield A&M",
                  "given_name": "Ian",
                  "surname": "Reynolds",
                },
              ],
            },
            "doi_data": Object {
              "doi": "an_utterly_fake_doi",
              "resource": "https://test.com",
              "timestamp": 1555009335577,
            },
            "edition_number": "3",
            "isbn": "this_is_an_isbn",
            "publication_date": Object {
              "@media_type": "online",
              "day": "02",
              "month": "03",
              "year": "2015",
            },
            "publisher": Object {
              "publisher_name": "PubPub",
            },
            "titles": Object {
              "title": "My new book full of pubs",
            },
          },
        },
      },
      "head": Object {
        "depositor": Object {
          "depositor_name": "PubPub",
          "email_address": "pubpub@media.mit.edu",
        },
        "doi_batch_id": "1555009335577_eea8ec7d",
        "registrant": "PubPub",
        "timestamp": 1555009335577,
      },
    },
  },
  "dois": Object {
    "collection": "an_utterly_fake_doi",
    "community": "10.21428/eea8ec7d",
    "pub": undefined,
  },
  "timestamp": 1555009335577,
}
`;

exports[`createDeposit creates a deposit for a book without any date information 1`] = `
Object {
  "deposit": Object {
    "doi_batch": Object {
      "@version": "4.4.1",
      "@xmlns": "http://www.crossref.org/schema/4.4.1",
      "@xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
      "@xsi:schemaLocation": "http://www.crossref.org/schema/4.4.1 http://www.crossref.org/schema/deposit/crossref4.4.1.xsd",
      "body": Object {
        "book": Object {
          "@book_type": "other",
          "book_metadata": Object {
            "contributors": Object {
              "person_name": Array [
                Object {
                  "@contributor_role": "author",
                  "@sequence": "first",
                  "affiliation": "Massachusetts Institute of Technology",
                  "given_name": "Travis",
                  "surname": "Rich",
                },
                Object {
                  "@contributor_role": "reader",
                  "@sequence": "additional",
                  "given_name": "Gabriel",
                  "surname": "Stein",
                },
                Object {
                  "@contributor_role": "reader",
                  "@sequence": "additional",
                  "given_name": "Catherine",
                  "surname": "Ahearn",
                },
                Object {
                  "@contributor_role": "author",
                  "@sequence": "additional",
                  "affiliation": "Springfield A&M",
                  "given_name": "Ian",
                  "surname": "Reynolds",
                },
              ],
            },
            "doi_data": Object {
              "doi": "an_utterly_fake_doi",
              "resource": "https://test.com",
              "timestamp": 1555009335577,
            },
            "edition_number": "3",
            "isbn": "this_is_an_isbn",
            "publication_date": Object {
              "@media_type": "online",
              "day": "11",
              "month": "04",
              "year": "2019",
            },
            "publisher": Object {
              "publisher_name": "PubPub",
            },
            "titles": Object {
              "title": "My new book full of pubs",
            },
          },
        },
      },
      "head": Object {
        "depositor": Object {
          "depositor_name": "PubPub",
          "email_address": "pubpub@media.mit.edu",
        },
        "doi_batch_id": "1555009335577_eea8ec7d",
        "registrant": "PubPub",
        "timestamp": 1555009335577,
      },
    },
  },
  "dois": Object {
    "collection": "an_utterly_fake_doi",
    "community": "10.21428/eea8ec7d",
    "pub": undefined,
  },
  "timestamp": 1555009335577,
}
`;

exports[`createDeposit creates a deposit for a conference 1`] = `
Object {
  "deposit": Object {
    "doi_batch": Object {
      "@version": "4.4.1",
      "@xmlns": "http://www.crossref.org/schema/4.4.1",
      "@xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
      "@xsi:schemaLocation": "http://www.crossref.org/schema/4.4.1 http://www.crossref.org/schema/deposit/crossref4.4.1.xsd",
      "body": Object {
        "conference": Object {
          "contributors": Object {
            "person_name": Array [
              Object {
                "@contributor_role": "author",
                "@sequence": "first",
                "affiliation": "Massachusetts Institute of Technology",
                "given_name": "Travis",
                "surname": "Rich",
              },
              Object {
                "@contributor_role": "reader",
                "@sequence": "additional",
                "given_name": "Gabriel",
                "surname": "Stein",
              },
              Object {
                "@contributor_role": "reader",
                "@sequence": "additional",
                "given_name": "Catherine",
                "surname": "Ahearn",
              },
              Object {
                "@contributor_role": "author",
                "@sequence": "additional",
                "affiliation": "Springfield A&M",
                "given_name": "Ian",
                "surname": "Reynolds",
              },
            ],
          },
          "event_metadata": Object {
            "conference_acronym": "TCIHA",
            "conference_date": Object {
              "#text": "2019-05-01",
            },
            "conference_location": "Kalamazoo, MI",
            "conference_name": "This Conference Is Happening Again",
            "conference_theme": "We love conferences don't we",
          },
          "proceedings_metadata": Object {
            "doi_data": Object {
              "doi": "yes_it_is_a_conference_doi",
              "resource": "https://dev.pubpub.org/collection/4c3a0515",
              "timestamp": 1555009335577,
            },
            "noisbn": Object {
              "@reason": "archive_volume",
            },
            "proceedings_title": "This Conference Is Happening Again",
            "publication_date": Object {
              "@media_type": "online",
              "day": "01",
              "month": "05",
              "year": "2019",
            },
            "publisher": Object {
              "publisher_name": "PubPub",
            },
          },
        },
      },
      "head": Object {
        "depositor": Object {
          "depositor_name": "PubPub",
          "email_address": "pubpub@media.mit.edu",
        },
        "doi_batch_id": "1555009335577_eea8ec7d",
        "registrant": "PubPub",
        "timestamp": 1555009335577,
      },
    },
  },
  "dois": Object {
    "collection": "yes_it_is_a_conference_doi",
    "community": "10.21428/eea8ec7d",
    "pub": undefined,
  },
  "timestamp": 1555009335577,
}
`;

exports[`createDeposit creates a deposit for a conference without any date information 1`] = `
Object {
  "deposit": Object {
    "doi_batch": Object {
      "@version": "4.4.1",
      "@xmlns": "http://www.crossref.org/schema/4.4.1",
      "@xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
      "@xsi:schemaLocation": "http://www.crossref.org/schema/4.4.1 http://www.crossref.org/schema/deposit/crossref4.4.1.xsd",
      "body": Object {
        "conference": Object {
          "contributors": Object {
            "person_name": Array [
              Object {
                "@contributor_role": "author",
                "@sequence": "first",
                "affiliation": "Massachusetts Institute of Technology",
                "given_name": "Travis",
                "surname": "Rich",
              },
              Object {
                "@contributor_role": "reader",
                "@sequence": "additional",
                "given_name": "Gabriel",
                "surname": "Stein",
              },
              Object {
                "@contributor_role": "reader",
                "@sequence": "additional",
                "given_name": "Catherine",
                "surname": "Ahearn",
              },
              Object {
                "@contributor_role": "author",
                "@sequence": "additional",
                "affiliation": "Springfield A&M",
                "given_name": "Ian",
                "surname": "Reynolds",
              },
            ],
          },
          "event_metadata": Object {
            "conference_acronym": "TCIHA",
            "conference_date": Object {
              "#text": "2019-04-11",
            },
            "conference_location": "Kalamazoo, MI",
            "conference_name": "This Conference Is Happening Again",
            "conference_theme": "We love conferences don't we",
          },
          "proceedings_metadata": Object {
            "doi_data": Object {
              "doi": "yes_it_is_a_conference_doi",
              "resource": "https://dev.pubpub.org/collection/4c3a0515",
              "timestamp": 1555009335577,
            },
            "noisbn": Object {
              "@reason": "archive_volume",
            },
            "proceedings_title": "This Conference Is Happening Again",
            "publication_date": Object {
              "@media_type": "online",
              "day": "11",
              "month": "04",
              "year": "2019",
            },
            "publisher": Object {
              "publisher_name": "PubPub",
            },
          },
        },
      },
      "head": Object {
        "depositor": Object {
          "depositor_name": "PubPub",
          "email_address": "pubpub@media.mit.edu",
        },
        "doi_batch_id": "1555009335577_eea8ec7d",
        "registrant": "PubPub",
        "timestamp": 1555009335577,
      },
    },
  },
  "dois": Object {
    "collection": "yes_it_is_a_conference_doi",
    "community": "10.21428/eea8ec7d",
    "pub": undefined,
  },
  "timestamp": 1555009335577,
}
`;

exports[`createDeposit creates a deposit for a pub 1`] = `
Object {
  "deposit": Object {
    "doi_batch": Object {
      "@version": "4.4.1",
      "@xmlns": "http://www.crossref.org/schema/4.4.1",
      "@xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
      "@xsi:schemaLocation": "http://www.crossref.org/schema/4.4.1 http://www.crossref.org/schema/deposit/crossref4.4.1.xsd",
      "body": Object {
        "journal": Object {
          "journal_article": Object {
            "@publication_type": "full_text",
            "contributors": Object {
              "person_name": Array [
                Object {
                  "@contributor_role": "author",
                  "@sequence": "first",
                  "affiliation": "Springfield A&M",
                  "given_name": "Ian",
                  "surname": "Reynolds",
                },
                Object {
                  "@contributor_role": "reader",
                  "@sequence": "additional",
                  "given_name": "Kevin",
                  "surname": "Bentham",
                },
              ],
            },
            "doi_data": Object {
              "doi": "10.21428/eea8ec7d.1519656b",
              "resource": "https://frankdev.pubpub.org/pub/h2muzdpv",
              "timestamp": 1555009335577,
            },
            "publication_date": Object {
              "@media_type": "online",
              "day": "01",
              "month": "03",
              "year": "2019",
            },
            "titles": Object {
              "title": "In Which Ian Makes A Pub",
            },
          },
          "journal_metadata": Object {
            "@language": "en",
            "abbrev_title": "PubPub Dev",
            "doi_data": Object {
              "doi": "10.21428/eea8ec7d",
              "resource": "https://frankdev.pubpub.org",
              "timestamp": 1555009335577,
            },
            "full_title": "PubPub Dev",
          },
        },
      },
      "head": Object {
        "depositor": Object {
          "depositor_name": "PubPub",
          "email_address": "pubpub@media.mit.edu",
        },
        "doi_batch_id": "1555009335577_eea8ec7d",
        "registrant": "PubPub",
        "timestamp": 1555009335577,
      },
    },
  },
  "dois": Object {
    "collection": undefined,
    "community": "10.21428/eea8ec7d",
    "pub": "10.21428/eea8ec7d.1519656b",
  },
  "timestamp": 1555009335577,
}
`;

exports[`createDeposit creates a deposit for a pub in a book 1`] = `
Object {
  "deposit": Object {
    "doi_batch": Object {
      "@version": "4.4.1",
      "@xmlns": "http://www.crossref.org/schema/4.4.1",
      "@xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
      "@xsi:schemaLocation": "http://www.crossref.org/schema/4.4.1 http://www.crossref.org/schema/deposit/crossref4.4.1.xsd",
      "body": Object {
        "book": Object {
          "@book_type": "other",
          "book_metadata": Object {
            "contributors": Object {
              "person_name": Array [
                Object {
                  "@contributor_role": "author",
                  "@sequence": "first",
                  "affiliation": "Massachusetts Institute of Technology",
                  "given_name": "Travis",
                  "surname": "Rich",
                },
                Object {
                  "@contributor_role": "reader",
                  "@sequence": "additional",
                  "given_name": "Gabriel",
                  "surname": "Stein",
                },
                Object {
                  "@contributor_role": "reader",
                  "@sequence": "additional",
                  "given_name": "Catherine",
                  "surname": "Ahearn",
                },
                Object {
                  "@contributor_role": "author",
                  "@sequence": "additional",
                  "affiliation": "Springfield A&M",
                  "given_name": "Ian",
                  "surname": "Reynolds",
                },
              ],
            },
            "doi_data": Object {
              "doi": "an_utterly_fake_doi",
              "resource": "https://test.com",
              "timestamp": 1555009335577,
            },
            "edition_number": "3",
            "isbn": "this_is_an_isbn",
            "publication_date": Object {
              "@media_type": "online",
              "day": "02",
              "month": "03",
              "year": "2015",
            },
            "publisher": Object {
              "publisher_name": "PubPub",
            },
            "titles": Object {
              "title": "My new book full of pubs",
            },
          },
          "content_item": Object {
            "@component_type": "chapter",
            "contributors": Object {
              "person_name": Array [
                Object {
                  "@contributor_role": "author",
                  "@sequence": "first",
                  "affiliation": "Springfield A&M",
                  "given_name": "Ian",
                  "surname": "Reynolds",
                },
                Object {
                  "@contributor_role": "reader",
                  "@sequence": "additional",
                  "given_name": "Kevin",
                  "surname": "Bentham",
                },
              ],
            },
            "doi_data": Object {
              "doi": "10.21428/eea8ec7d.1519656b",
              "resource": "https://frankdev.pubpub.org/pub/h2muzdpv",
              "timestamp": 1555009335577,
            },
            "publication_date": Object {
              "@media_type": "online",
              "day": "01",
              "month": "03",
              "year": "2019",
            },
            "titles": Object {
              "title": "In Which Ian Makes A Pub",
            },
          },
        },
      },
      "head": Object {
        "depositor": Object {
          "depositor_name": "PubPub",
          "email_address": "pubpub@media.mit.edu",
        },
        "doi_batch_id": "1555009335577_eea8ec7d",
        "registrant": "PubPub",
        "timestamp": 1555009335577,
      },
    },
  },
  "dois": Object {
    "collection": "an_utterly_fake_doi",
    "community": "10.21428/eea8ec7d",
    "pub": "10.21428/eea8ec7d.1519656b",
  },
  "timestamp": 1555009335577,
}
`;

exports[`createDeposit creates a deposit for a pub in a conference 1`] = `
Object {
  "deposit": Object {
    "doi_batch": Object {
      "@version": "4.4.1",
      "@xmlns": "http://www.crossref.org/schema/4.4.1",
      "@xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
      "@xsi:schemaLocation": "http://www.crossref.org/schema/4.4.1 http://www.crossref.org/schema/deposit/crossref4.4.1.xsd",
      "body": Object {
        "conference": Object {
          "conference_paper": Object {
            "contributors": Object {
              "person_name": Array [
                Object {
                  "@contributor_role": "author",
                  "@sequence": "first",
                  "affiliation": "Springfield A&M",
                  "given_name": "Ian",
                  "surname": "Reynolds",
                },
                Object {
                  "@contributor_role": "reader",
                  "@sequence": "additional",
                  "given_name": "Kevin",
                  "surname": "Bentham",
                },
              ],
            },
            "doi_data": Object {
              "doi": "10.21428/eea8ec7d.1519656b",
              "resource": "https://frankdev.pubpub.org/pub/h2muzdpv",
              "timestamp": 1555009335577,
            },
            "titles": Object {
              "title": "In Which Ian Makes A Pub",
            },
          },
          "contributors": Object {
            "person_name": Array [
              Object {
                "@contributor_role": "author",
                "@sequence": "first",
                "affiliation": "Massachusetts Institute of Technology",
                "given_name": "Travis",
                "surname": "Rich",
              },
              Object {
                "@contributor_role": "reader",
                "@sequence": "additional",
                "given_name": "Gabriel",
                "surname": "Stein",
              },
              Object {
                "@contributor_role": "reader",
                "@sequence": "additional",
                "given_name": "Catherine",
                "surname": "Ahearn",
              },
              Object {
                "@contributor_role": "author",
                "@sequence": "additional",
                "affiliation": "Springfield A&M",
                "given_name": "Ian",
                "surname": "Reynolds",
              },
            ],
          },
          "event_metadata": Object {
            "conference_acronym": "TCIHA",
            "conference_date": Object {
              "#text": "2019-05-01",
            },
            "conference_location": "Kalamazoo, MI",
            "conference_name": "This Conference Is Happening Again",
            "conference_theme": "We love conferences don't we",
          },
          "proceedings_metadata": Object {
            "doi_data": Object {
              "doi": "yes_it_is_a_conference_doi",
              "resource": "https://dev.pubpub.org/collection/4c3a0515",
              "timestamp": 1555009335577,
            },
            "noisbn": Object {
              "@reason": "archive_volume",
            },
            "proceedings_title": "This Conference Is Happening Again",
            "publication_date": Object {
              "@media_type": "online",
              "day": "01",
              "month": "05",
              "year": "2019",
            },
            "publisher": Object {
              "publisher_name": "PubPub",
            },
          },
        },
      },
      "head": Object {
        "depositor": Object {
          "depositor_name": "PubPub",
          "email_address": "pubpub@media.mit.edu",
        },
        "doi_batch_id": "1555009335577_eea8ec7d",
        "registrant": "PubPub",
        "timestamp": 1555009335577,
      },
    },
  },
  "dois": Object {
    "collection": "yes_it_is_a_conference_doi",
    "community": "10.21428/eea8ec7d",
    "pub": "10.21428/eea8ec7d.1519656b",
  },
  "timestamp": 1555009335577,
}
`;

exports[`createDeposit creates a deposit for a pub in an issue 1`] = `
Object {
  "deposit": Object {
    "doi_batch": Object {
      "@version": "4.4.1",
      "@xmlns": "http://www.crossref.org/schema/4.4.1",
      "@xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
      "@xsi:schemaLocation": "http://www.crossref.org/schema/4.4.1 http://www.crossref.org/schema/deposit/crossref4.4.1.xsd",
      "body": Object {
        "journal": Object {
          "journal_article": Object {
            "@publication_type": "full_text",
            "contributors": Object {
              "person_name": Array [
                Object {
                  "@contributor_role": "author",
                  "@sequence": "first",
                  "affiliation": "Springfield A&M",
                  "given_name": "Ian",
                  "surname": "Reynolds",
                },
                Object {
                  "@contributor_role": "reader",
                  "@sequence": "additional",
                  "given_name": "Kevin",
                  "surname": "Bentham",
                },
              ],
            },
            "doi_data": Object {
              "doi": "10.21428/eea8ec7d.1519656b",
              "resource": "https://frankdev.pubpub.org/pub/h2muzdpv",
              "timestamp": 1555009335577,
            },
            "publication_date": Object {
              "@media_type": "online",
              "day": "01",
              "month": "03",
              "year": "2019",
            },
            "titles": Object {
              "title": "In Which Ian Makes A Pub",
            },
          },
          "journal_issue": Object {
            "contributors": Object {
              "person_name": Array [
                Object {
                  "@contributor_role": "author",
                  "@sequence": "first",
                  "affiliation": "Massachusetts Institute of Technology",
                  "given_name": "Travis",
                  "surname": "Rich",
                },
                Object {
                  "@contributor_role": "reader",
                  "@sequence": "additional",
                  "given_name": "Gabriel",
                  "surname": "Stein",
                },
                Object {
                  "@contributor_role": "reader",
                  "@sequence": "additional",
                  "given_name": "Catherine",
                  "surname": "Ahearn",
                },
                Object {
                  "@contributor_role": "author",
                  "@sequence": "additional",
                  "affiliation": "Springfield A&M",
                  "given_name": "Ian",
                  "surname": "Reynolds",
                },
              ],
            },
            "doi_data": Object {
              "doi": "10.21428/eea8ec7d.e281f63f",
              "resource": "https://dev.pubpub.org/collection/e281f63f",
              "timestamp": 1555009335577,
            },
            "issue": "5",
            "journal_volume": Object {
              "volume": "2",
            },
            "publication_date": Object {
              "@media_type": "online",
              "day": "25",
              "month": "03",
              "year": "2017",
            },
            "titles": Object {
              "title": "Another Great Issue",
            },
          },
          "journal_metadata": Object {
            "@language": "en",
            "abbrev_title": "PubPub Dev",
            "doi_data": Object {
              "doi": "10.21428/eea8ec7d",
              "resource": "https://frankdev.pubpub.org",
              "timestamp": 1555009335577,
            },
            "full_title": "PubPub Dev",
          },
        },
      },
      "head": Object {
        "depositor": Object {
          "depositor_name": "PubPub",
          "email_address": "pubpub@media.mit.edu",
        },
        "doi_batch_id": "1555009335577_eea8ec7d",
        "registrant": "PubPub",
        "timestamp": 1555009335577,
      },
    },
  },
  "dois": Object {
    "collection": "10.21428/eea8ec7d.e281f63f",
    "community": "10.21428/eea8ec7d",
    "pub": "10.21428/eea8ec7d.1519656b",
  },
  "timestamp": 1555009335577,
}
`;

exports[`createDeposit creates a deposit for an issue 1`] = `
Object {
  "deposit": Object {
    "doi_batch": Object {
      "@version": "4.4.1",
      "@xmlns": "http://www.crossref.org/schema/4.4.1",
      "@xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
      "@xsi:schemaLocation": "http://www.crossref.org/schema/4.4.1 http://www.crossref.org/schema/deposit/crossref4.4.1.xsd",
      "body": Object {
        "journal": Object {
          "journal_issue": Object {
            "contributors": Object {
              "person_name": Array [
                Object {
                  "@contributor_role": "author",
                  "@sequence": "first",
                  "affiliation": "Massachusetts Institute of Technology",
                  "given_name": "Travis",
                  "surname": "Rich",
                },
                Object {
                  "@contributor_role": "reader",
                  "@sequence": "additional",
                  "given_name": "Gabriel",
                  "surname": "Stein",
                },
                Object {
                  "@contributor_role": "reader",
                  "@sequence": "additional",
                  "given_name": "Catherine",
                  "surname": "Ahearn",
                },
                Object {
                  "@contributor_role": "author",
                  "@sequence": "additional",
                  "affiliation": "Springfield A&M",
                  "given_name": "Ian",
                  "surname": "Reynolds",
                },
              ],
            },
            "doi_data": Object {
              "doi": "10.21428/eea8ec7d.e281f63f",
              "resource": "https://dev.pubpub.org/collection/e281f63f",
              "timestamp": 1555009335577,
            },
            "issue": "5",
            "journal_volume": Object {
              "volume": "2",
            },
            "publication_date": Object {
              "@media_type": "online",
              "day": "25",
              "month": "03",
              "year": "2017",
            },
            "titles": Object {
              "title": "Another Great Issue",
            },
          },
          "journal_metadata": Object {
            "@language": "en",
            "abbrev_title": "PubPub Dev",
            "doi_data": Object {
              "doi": "10.21428/eea8ec7d",
              "resource": "https://frankdev.pubpub.org",
              "timestamp": 1555009335577,
            },
            "full_title": "PubPub Dev",
          },
        },
      },
      "head": Object {
        "depositor": Object {
          "depositor_name": "PubPub",
          "email_address": "pubpub@media.mit.edu",
        },
        "doi_batch_id": "1555009335577_eea8ec7d",
        "registrant": "PubPub",
        "timestamp": 1555009335577,
      },
    },
  },
  "dois": Object {
    "collection": "10.21428/eea8ec7d.e281f63f",
    "community": "10.21428/eea8ec7d",
    "pub": undefined,
  },
  "timestamp": 1555009335577,
}
`;

exports[`createDeposit creates a deposit for an issue without any date information 1`] = `
Object {
  "deposit": Object {
    "doi_batch": Object {
      "@version": "4.4.1",
      "@xmlns": "http://www.crossref.org/schema/4.4.1",
      "@xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
      "@xsi:schemaLocation": "http://www.crossref.org/schema/4.4.1 http://www.crossref.org/schema/deposit/crossref4.4.1.xsd",
      "body": Object {
        "journal": Object {
          "journal_issue": Object {
            "contributors": Object {
              "person_name": Array [
                Object {
                  "@contributor_role": "author",
                  "@sequence": "first",
                  "affiliation": "Massachusetts Institute of Technology",
                  "given_name": "Travis",
                  "surname": "Rich",
                },
                Object {
                  "@contributor_role": "reader",
                  "@sequence": "additional",
                  "given_name": "Gabriel",
                  "surname": "Stein",
                },
                Object {
                  "@contributor_role": "reader",
                  "@sequence": "additional",
                  "given_name": "Catherine",
                  "surname": "Ahearn",
                },
                Object {
                  "@contributor_role": "author",
                  "@sequence": "additional",
                  "affiliation": "Springfield A&M",
                  "given_name": "Ian",
                  "surname": "Reynolds",
                },
              ],
            },
            "doi_data": Object {
              "doi": "10.21428/eea8ec7d.e281f63f",
              "resource": "https://dev.pubpub.org/collection/e281f63f",
              "timestamp": 1555009335577,
            },
            "issue": "5",
            "journal_volume": Object {
              "volume": "2",
            },
            "publication_date": Object {
              "@media_type": "online",
              "day": "11",
              "month": "04",
              "year": "2019",
            },
            "titles": Object {
              "title": "Another Great Issue",
            },
          },
          "journal_metadata": Object {
            "@language": "en",
            "abbrev_title": "PubPub Dev",
            "doi_data": Object {
              "doi": "10.21428/eea8ec7d",
              "resource": "https://frankdev.pubpub.org",
              "timestamp": 1555009335577,
            },
            "full_title": "PubPub Dev",
          },
        },
      },
      "head": Object {
        "depositor": Object {
          "depositor_name": "PubPub",
          "email_address": "pubpub@media.mit.edu",
        },
        "doi_batch_id": "1555009335577_eea8ec7d",
        "registrant": "PubPub",
        "timestamp": 1555009335577,
      },
    },
  },
  "dois": Object {
    "collection": "10.21428/eea8ec7d.e281f63f",
    "community": "10.21428/eea8ec7d",
    "pub": undefined,
  },
  "timestamp": 1555009335577,
}
`;
